### ISSUE TITLE
separate dense and free

### ISSUE THREAD

**Issue by NicoRenaud** [NLESC-JCER/Fortran_Davidson#40]
Should we separate the dense and free part of the code ? Having everything in one file makes it difficult to navigate (at least for me).

Also some subroutines have a dense and a free version while I think they should be the same (for example the DPR correction). We should simplify that.

**Comment by felipeZ** [comment]
I agree about separating the two implementations. Also, I have tried to make generic functions like DPR correction but I have not succeeded

**Comment by NicoRenaud** [comment]
I can look into the generic DPR. At the end of the day we only need the residues and the diagonal. So nothing of that require a free mat vec product

**Comment by felipeZ** [comment]
I Just realized that once we removed the `ritz_vectors` recalculation, the DPR free and dense subroutines are basically the same
